saleor/graphql/product/mutations/channels.py

```diff
@@ -0,0 +1,185 @@
+from collections import defaultdict
+from typing import TYPE_CHECKING, DefaultDict, Dict, Iterable, List
+
+import graphene
+from django.core.exceptions import ValidationError
+from django.db import transaction
+
+from ....core.permissions import ProductPermissions
+from ....product.error_codes import ProductErrorCode
+from ....product.models import ProductChannelListing
+from ...channel.types import Channel
+from ...core.mutations import BaseMutation
+from ...core.types.common import ProductChannelListingError
+from ...core.utils import get_duplicated_values, get_duplicates_ids
+from ...utils import resolve_global_ids_to_primary_keys
+from ..types.products import Product
+
+if TYPE_CHECKING:
+    from ....product.models import Product as ProductModel
+    from ....channel.models import Channel as ChannelModel
+
+ErrorType = DefaultDict[str, List[ValidationError]]
+
+
+class ProductChannelListingAddInput(graphene.InputObjectType):
+    channel_id = graphene.ID(
+        required=True, description="ID of a channel witch be assigned to product."
+    )
+    is_published = graphene.Boolean(
+        description="Determines if product is visible to customers.", required=True
+    )
+    publication_date = graphene.types.datetime.Date(
+        description="Publication date. ISO 8601 standard."
+    )
+
+
+class ProductChannelListingUpdateInput(graphene.InputObjectType):
+    add_channels = graphene.List(
+        graphene.NonNull(ProductChannelListingAddInput),
+        description="List of channel assigned to product.",
+        required=False,
+    )
+    remove_channels = graphene.List(
+        graphene.NonNull(graphene.ID),
+        description="List of channel unassigned from product.",
+        required=False,
+    )
+
+
+class ProductChannelListingUpdate(BaseMutation):
+    product = graphene.Field(Product, description="An updated product instance.")
+
+    class Arguments:
+        id = graphene.ID(required=True, description="ID of a product to update.")
+        input = ProductChannelListingUpdateInput(
+            required=True,
+            description="Fields required to create product channel listings.",
+        )
+
+    class Meta:
+        description = "Update product chanel listing."
```

Comment:
The "channel listing" term is fine for mutations and internals but in the public description/docs I would explain more clearly that it refers to creating/updating product availability in a channel - this is how people should understand that.

```suggestion
        description = "Manage product's availability in channels."
```